PRESIDECMS-3104 consume workflow extension

DominicWatson opened the pull request:
This PR takes Pixl8's private workflow extension and makes it a first class citizen in the Preside platform.

* Full data model migrated
* cfflow now a preside dependency
* webflow system added
* datamanager workflow system added and made first class citizen of datamanager
* feature flags added to control this functionality and enabled by default

DominicWatson committed:
[twgit] Init feature 'feature-PRESIDECMS-3104_consume-workflow-extension'.

DominicWatson committed:
PRESIDECMS-3104 Migrate workflow extension into core Preside

DominicWatson committed:
PRESIDECMS-3104 import test cases from workflow extension

Pixl8RusselCole commented on system/handlers/renderers/content/WebflowStep.cfc:
```diff
@@ -0,0 +1,25 @@
+	}
+
+	private string function adminDataTable( event, rc, prc, args={} ) {
+		var stepId = args.data ?: "";
```
Copied file I know, but can we align the equals signs

Pixl8RusselCole commented on system/services/datamanagerflow/DatamanagerWorkflowFilterService.cfc:
```diff
@@ -0,0 +1,211 @@
+			var subQuery         = $getPresideObject( "cfflow_workflow_instance_step" ).selectData(
+				  selectFields        = [ 1 ]
+				, filter              = { step=nonCompleteSteps, status="active", "instance.workflow_id"=cfflowWfId, "instance.reference"=arguments.objectName }
+				, extraFilters        = [ { filter=outerJoin }]
```
Alignment

Pixl8RusselCole commented on system/services/webflow/WebflowArchivingService.cfc:
```diff
@@ -0,0 +1,110 @@
+
+// PUBLIC API METHODS
+	public boolean function archiveExpiredWebflows( any logger ) {
+		var canLog       = StructKeyExists( arguments, "logger" );
```
Maybe change to arguments.logger?.
And remove the canInfo canDebug logic

Pixl8RusselCole commented on system/services/webflow/WebflowSubmissionService.cfc:
```diff
@@ -0,0 +1,358 @@
+			, instance  = local.wfInstance ?: NullValue()
+		);
+
+		var webflow = _getWebflowLibrary().getWebflow( webflowId );
```
Alignment

Pixl8RusselCole commented on system/services/webflow/WebflowSubmissionService.cfc:
```diff
@@ -0,0 +1,358 @@
+		if ( Len( preCancel.event ?: "" ) ) {
+			$runEvent(
+				event = preCancel.event
+				, private = true
```
Alignment

Pixl8RusselCole commented on system/services/webflow/WebflowSubmissionService.cfc:
```diff
@@ -0,0 +1,358 @@
+		if ( Len( postCancel.event ?: "" ) ) {
+			$runEvent(
+				event = postCancel.event
+				, private = true
```
Alignment

DominicWatson committed:
PRESIDECMS-3104 little fixes

* Add admin menu item
* Add legacy mapping to cfflow

DominicWatson committed:
PRESIDECMS-3104 lil tidy ups from code review

pixl8-brayden committed:
PRESIDECMS-3104 move latest webflow progress bar update from private repo

DominicWatson committed:
PRESIDECMS-3104 add documentation

DominicWatson committed:
PRESIDECMS-3104 WF-92 do not delete existing owner flows

When upgrading a session from visitor -> user. Preserve and prefer
existing user flow instances rather than transfering the new visitor
instances. This prevents issues with workflow data going missing
in various session timeout scenarios, etc. and is a much more sensible
and safe default.

sebduggan committed:
PRESIDECMS-3104 WF-92 Archive visitor flow if using existing userid flow

pixl8-brayden committed:
PRESIDECMS-3104 add nofollow for webflow prev button link
